File: content-solutions/stack/undo-memo.py
def solution(commands):
    text = ""
    history = []
    for command in commands:
        parts = command.split()
        name = parts[0]
        if name == "type":
            history.append(text)
            text += parts[1]
        elif name == "delete":
            history.append(text)
            text = text[: max(0, len(text) - int(parts[1]))]
        elif history:
            text = history.pop()
    return text

File: content-solutions/stack/undo-memo.js
function solution(commands) {
  let text = "";
  const history = [];
  for (const command of commands) {
    const [name, arg] = command.split(" ");
    if (name === "type") {
      history.push(text);
      text += arg;
    } else if (name === "delete") {
      history.push(text);
      text = text.slice(0, Math.max(0, text.length - Number(arg)));
    } else if (history.length > 0) {
      text = history.pop();
    }
  }
  return text;
}
